pkg/kconfig/minimize.go

```diff
@@ -4,79 +4,65 @@
-	// If it does not crash, try another half.
-	// If the crash is caused by a single config, this algorithm is guaranteed to find it.
-	// If the crash is caused by multiple configs, this algorithm will most likely find them (along with some
-	// additional unrelated configs that happened to be in the same half). However, amount of unrelated configs
-	// can be quite large if we are unlucky. Also note that we sort configs so that related configs are most
-	// likely situated together.
-	// Numerous improvements are possible for this simple algorithm.
-	// 1. We could split the config onto 4 parts and try all pairs, this should find all pairs of configs reliably.
-	// 2. We could continue trying to reduce a part even if removing the whole part fails. I.e. we try to remove
-	//    a half and it fails, we can try to remove half of the half, maybe that will succeed.
-top:
-	for len(diff) >= 2 {
-		half := len(diff) / 2
-		for _, part := range [][]string{diff[:half], diff[half:]} {
-			dt.Log("trying half: %v", part)
-			closure := kconf.addDependencies(base, full, part)
-			candidate := base.clone()
-			// Always move all non-tristate configs from full to base as we don't minimize them.
-			for _, cfg := range other {
-				candidate.Set(cfg.Name, cfg.Value)
-			}
-			for _, cfg := range closure {
-				candidate.Set(cfg, Yes)
-			}
-			res, err := pred(candidate)
-			if err != nil {
-				return nil, err
-			}
-			if res {
-				dt.Log("half crashed")
-				diff = part
-				current = candidate
-				suspects = closure
-				continue top
-			}
+	maxSteps int, dt debugtracer.DebugTracer) (*ConfigFile, error) {
+	diff, other := kconf.missingLeafConfigs(base, full)
+	dt.Log("kconfig minimization: base=%v full=%v leaves diff=%v", len(base.Configs), len(full.Configs), len(diff))
+
+	diffToConfig := func(part []string) (*ConfigFile, []string) {
+		if len(part) == 0 {
+			// We're testing the baseline config only.
+			return base, nil
+		}
+		suspects := kconf.addDependencies(base, full, part)
+		candidate := base.clone()
+		// Always move all non-tristate configs from full to base as we don't minimize them.
+		for _, cfg := range other {
+			candidate.Set(cfg.Name, cfg.Value)
 		}
-		dt.Log("both halves did not crash")
-		break
+		for _, cfg := range suspects {
+			candidate.Set(cfg, Yes)
+		}
+		return candidate, suspects
 	}
+	step := 1
+	result, err := generic.MinimizeSlice(generic.MinimizeConfig[string]{
+		Pred: func(diffs []string) (bool, error) {
+			config, _ := diffToConfig(diffs)
+			dt.SaveFile(fmt.Sprintf("step_%d.config", step), config.Serialize())
+			step++
+			return pred(config)
+		},
+		MaxSteps: maxSteps,
+		Logf:     dt.Log,
+	},
+		// Let's be realistic -- it's rather unlikely that the kernel will crash with
+		// baseline config or with just one extra config.
+		// So let's begin with 3 chunks.
+		diff[0:len(diff)/3], diff[len(diff)/3:2*len(diff)/3], diff[2*len(diff)/3:],
```

Comment:
This feels out-of-place here. This is effectively first step of minimization, but it happens here, while the rest happens insize of MinimizeSlice.
How important is this splitting by 3?
Should all config minimization steps split in 3 instead of 2? I.e. should it it be a config parameter?
Or should we always split in 3 (also for repro)?
Or should we simply always split in 2?
It would be nice if MinimizeSlice accept just a single slice and then do the rest.

Reply:
This also ignores the baseline config. So either this is wrong, or if it works w/o the baseline config, then maybe we don't need it at all?

Reply:
I think we still need some knob to determine the original split. The minimization process would of course work just fine without it, but if we know (or guess) the average size of the resulting diff, that can help us save some predicate runs. E.g. if there are on average 2 necessary configs, it doesn't make much sense to run with empty diff, and, if we split it in half, there's only a `0.5*0.5=0.25` chance that a half could be dropped. If we split in 3, the chance that a predicate run results in a chunk being dropped is `0.44`.

Maybe add some `InitialSplitting` option to the minimization config?

> This also ignores the baseline config.

Why? We bisect a set of diffs between baseline and full + we apply the selected diff subset to the baseline inside the predicate.

Reply:
> Why? We bisect a set of diffs between baseline and full + we apply the selected diff subset to the baseline inside the predicate.

I was wrong, ignore.

Reply:
> E.g. if there are on average 2 necessary configs, it doesn't make much sense to run with empty diff, and, if we split it in half, there's only a 0.5*0.5=0.25 chance that a half could be dropped.

Isn't it 0.5?
The cance that the a config is in particular chunk out of 2 is 0.5 (seconds is in the same as the first one).

> If we split in 3, the chance that a predicate run results in a chunk being dropped is 0.44.

Doesn't this apply to other steps as well if the 2 configs happen to be in the same chunk?

Doesn't this also apply to repro minimization? Each individual program did not crash, so presumably we are looking for 2+ programs.

Reply:
I've just done a simple benchmarking: 300 initial elements, and a limit on 5 iterations (= what we'll likely use for quite a long time during bisections). The varying elements are the number of guilty/needed elements and the number of chunks we split array to before doing the minimization.

|Test|Runs|Average remaining len|
|-|-|-|
|1_guilty_1_starting_chunks-36|200|55.18|
|1_guilty_2_starting_chunks-36|200|46.38|
|1_guilty_3_starting_chunks-36|200|43.75|
|1_guilty_4_starting_chunks-36|200|56.50|
|1_guilty_5_starting_chunks-36|200|60.00|
|1_guilty_6_starting_chunks-36|200|94.25|
|2_guilty_1_starting_chunks-36|200|156.6|
|2_guilty_2_starting_chunks-36|200|130.3|
|2_guilty_3_starting_chunks-36|200|121.1|
|2_guilty_4_starting_chunks-36|200|120.6|
|2_guilty_5_starting_chunks-36|200|110.4|
|2_guilty_6_starting_chunks-36|200|125.5|
|3_guilty_1_starting_chunks-36|200|203.5|
|3_guilty_2_starting_chunks-36|200|181.0|
|3_guilty_3_starting_chunks-36|200|162.2|
|3_guilty_4_starting_chunks-36|200|159.2|
|3_guilty_5_starting_chunks-36|200|148.5|
|3_guilty_6_starting_chunks-36|200|156.0|

The code:
```golang
func BenchmarkSplits(b *testing.B) {
	for _, guilty := range []int{1, 2, 3} {
		guilty := guilty
		for _, chunks := range []int{1, 2, 3, 4, 5, 6} {
			chunks := chunks
			b.Run(fmt.Sprintf("%d_guilty_%d_starting_chunks", guilty, chunks), func(b *testing.B) {
				var sum int
				for i := 0; i < b.N; i++ {
					sum += runMinimize(guilty, chunks)
				}
				b.ReportMetric(float64(sum)/float64(b.N), "remaining-elements")
			})
		}
	}
}

func runMinimize(guilty, initialChunks int) int {
	const size = 300
	const steps = 5

	r := rand.New(rand.NewSource(time.Now().UnixNano()))
	array := make([]int, size)
	for _, j := range r.Perm(size)[:guilty] {
		array[j] = 1
	}

	var chunks [][]int
	chunkLen := len(array) / initialChunks
	for i := 0; i < len(array); i += chunkLen {
		end := i + chunkLen
		if end > len(array) {
			end = len(array)
		}
		chunks = append(chunks, array[i:end])
	}

	ret, _ := MinimizeSlice(MinimizeConfig[int]{
		MaxSteps: steps,
		Pred: func(arr []int) (bool, error) {
			nonZero := 0
			for _, x := range arr {
				if x > 0 {
					nonZero++
				}
			}
			return nonZero == guilty, nil
		},
	}, chunks...)
	return len(ret)
}
```

Reply:
> Doesn't this also apply to repro minimization?

I think it applies as well, though we pay a much smaller price for every extra predicate run than during kernel bisections. If we agree on using custom split for kernel config bisections, we can set something similar in pkg/repro.

Reply:
In your test the 1-chunk case is penalized due to dropIndividualChunks, which does not do anything useful, but consumes 20% of tests. If we don't do dropIndividualChunks when we have only 1 chunk, then 1 chunk spolit is the best fot 1 guilty and comparable for 2/3 guilty.

Another observation is that if we very few tests (5) then bisection does not really happen. Let's say we have 3 initial hunks, 3 tests consumed by dropIndividualChunks and then we are left 2 test for the whole bisection process.
What works best with so few tests is just dropIndividualChunks. I.e. if we have 5 tests, we split the set into 5 parts, test each one individually. No bisection. See your 2_guilty_5_starting_chunks and 3_guilty_5_starting_chunks.
So if we aim at 5 tests total, maybe we do just that?

Reply:
And please add the benchmark to the PR.

Reply:
> Another observation is that if we very few tests (5) then bisection does not really happen.

If the number of initial chunks is greater or equal than the number of max steps, bisection itself doesn't happen indeed.

> In your test the 1-chunk case is penalized due to dropIndividualChunks, which does not do anything useful

It covers/excludes the case when there are 0 guilty elements (i.e. minimization should return empty slice). And, as I mentioned above, we do need to first ensure that all chunks contain at least one needed element and then maintain this invariant throughout the process. It helps us save predicate executions.

That was one of the reasons to let users specify chunks -- if the caller has some expectations on the number of elements that must remain, it can make a better use of predicate runs.

> So if we aim at 5 tests total, maybe we do just that?

In the current circumstances, it seems it does make sense.

But definitely not in the generic case: see the runs with 1 guilty element -- 43/300, that's ~ 1/7, not 1/5. Once we start to do full config bisection, with an increased `MaxSteps` limit we'll see the same benefit from the actual bisection even for cases where there are more than one guilty element.

Wdyt about adding some `InitialSplit` parameter to `minimize.Config`? We shouldn't always use MaxSteps for this purpose for the reasons above, but we can use it to determine both parameters where we call the algorithm.

```
result, err := minimize.Slice(
		minimize.Config[string]{
			Pred:     minimizePred,
			MaxSteps: maxSteps,
			InitialSplit: min(maxSteps,  5),
			Logf:     dt.Log,
		}, diff)
```

Reply:
You refered to "expected number of elements that must remain" several times. It looks like that is the inherent characteristic users have in mind and should supply as an optimization.

How the split on the fisrt/subseuent steps is done is up to the minimization logic itself. It seems that it may benefit subsequent steps as well. E.g. if after initail step, we still have 1 chunk, it seems that it shuold be split in 3 parts again, no?

If callers supply the expected number, then they also don't need to do "min(maxSteps,  5)". The expected number may be larger than maxSteps. There is no contradiction here.

If for ~2 guilty configs, 3-way split works best, don't we want to do the same for repro minimization? We know that we are looking for >1 program. So it looks like 3 way split should work best as well.
So perhaps we don't specify this in the config, but assume that we are optimizing for >1 guilty elements and do 3-way split while we have 1 chunk?